ocean_lib/common/ddo/credentials.py

```diff
@@ -0,0 +1,106 @@
+        entries = self.asset._credentials.get(access_class, [])
+
+        for entry in entries:
+            if entry["type"] == "address":
+                return [val.lower() for val in entry["values"]]
+
+        return []
+
+    def requires_credential(self):
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        return allowed_addresses or denied_addresses
+
+    def validate_access(self, credential=None):
+        address = simplify_credential_to_address(credential)
+
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        if not address and not self.requires_credential():
+            return ConsumableCodes.OK
+
+        if allowed_addresses and address.lower() not in allowed_addresses:
+            return ConsumableCodes.CREDENTIAL_NOT_IN_ALLOW_LIST
+
+        if not allowed_addresses and address.lower() in denied_addresses:
+            return ConsumableCodes.CREDENTIAL_IN_DENY_LIST
+
+        return ConsumableCodes.OK
+
+    def add_address_to_list_class(self, address, list_class="allow"):
+        address = address.lower()
+
+        if not self.asset._credentials or list_class not in self.asset._credentials:
+            self.asset._credentials[list_class] = [
+                {"type": "address", "values": [address]}
+            ]
+            return
+
+        address_entry = get_list_entry_with_type_address(
+            self.asset._credentials[list_class]
+        )
+
+        if not address_entry:
+            self.asset._credentials[list_class].append(
+                {"type": "address", "values": [address]}
+            )
+            return
+
+        address_entry = address_entry[0]
+        lc_addresses = [addr.lower() for addr in address_entry["values"]]
+
+        if address not in lc_addresses:
+            lc_addresses.append(address)
+
+        address_entry["values"] = lc_addresses
+
+    def remove_address_from_list_class(self, address, list_class="allow"):
+        address = address.lower()
+
+        if not self.asset._credentials or list_class not in self.asset._credentials:
+            return
+
+        address_entry = get_list_entry_with_type_address(
+            self.asset._credentials[list_class]
+        )
+
+        if not address_entry:
+            return
+
+        address_entry = address_entry[0]
+        lc_addresses = [addr.lower() for addr in address_entry["values"]]
+
+        if address not in lc_addresses:
+            return
+
+        lc_addresses.remove(address)
+        address_entry["values"] = lc_addresses
+
+
+def simplify_credential_to_address(credential):
+    if not credential:
+        return None
+
+    if not credential.get("value"):
+        raise MalformedCredential("Received empty address.")
+
+    return credential["value"]
+
+
+def get_list_entry_with_type_address(types_list):
```

Comment:
I think this should be a class method instead of a module function. And it should match the conventions set by `get_addresses_of_class`.  In fact, `get_addresses_of_class` should use this.

```python
    def get_address_entry_of_class(self, access_class="allow"):
        """Get a filtered list of addresses of a given access_class from credentials (use with allow/deny)."""
        entries = self.asset._credentials.get(access_class, [])
        return [entry for entry in entries if entry["type"] == "address"]

    def get_addresses_of_class(self, access_class="allow"):
        """Get an address entry of a given access class from credentials (use with allow/deny)."""
        address_entry = get_address_entry_of_class(access_class)
        return [val.lower() for val in address_entry["values"]]
```

Reply:
My apologies, I just realized my previous comment mixed up the docstrings.

Reply:
I'll just resolve this comment because I think it's become stale compared to the latest commits.